File: java-algorithm/src/main/java/com/leetcode/Problem611.java
package com.leetcode;

import java.util.Arrays;

/**
 * 611. Valid Triangle Number
 * https://leetcode.com/problems/valid-triangle-number/
 */
public class Problem611 {
    public int triangleNumber(int[] nums) {
        int n = nums.length;
        if (n < 3) return 0;
        int ans = 0;
        Arrays.sort(nums);
        for (int i = 0; i < n - 2; i++) {
            for (int j = i + 1; j < n - 1; j++) {
                for (int k = j + 1; k < n; k++) {
                    if (nums[i] + nums[j] > nums[k])
                        ans++;
                    else
                        break;
                }
            }
        }
        return ans;
    }
}

File: python-algo/leetcode/problem_611.py
"""611. Valid Triangle Number
https://leetcode.com/problems/valid-triangle-number/
"""
import bisect
from typing import List


def triangle_number(nums: List[int]) -> int:
    n = len(nums)
    if n < 3:
        return 0
    nums.sort()
    ans = 0
    for i in range(n - 2):
        for j in range(i + 1, n - 1):
            val = nums[i] + nums[j]
            idx = bisect.bisect_left(nums, val)
            if idx == n:
                idx -= 1
            if nums[idx] >= val:
                idx -= 1
            if idx > j:
                ans += idx - j
    return ans
